server/src/main/java/org/elasticsearch/script/TimeSeriesCounter.java

```diff
@@ -0,0 +1,399 @@
+         *
+         *                    `nextBucketStartTime()`[d] is now 340, the start time of bucket 2.
+         *
+         *                    `earliestTimeInCounter()`[e] is now 240 as bucket 1 now represents 320-339 rather than 220-239.
+         *
+         *                    `counterExpired()`[f] is now 420.  Bucket 1 in the current epoch represents 320-339,  an
+         *                    `inc(long)` any of time (320 + `duration`) or beyond would require clearing all `buckets` in the
+         *                    `Counter` and any `sum(long)` that starts at 420 or later does not cover the valid time range for
+         *                    this state of the counter.
+         *
+         *                    `nextBucket(curBucket)`[g] is now 2, the bucket after 1.
+         *
+         * Past
+         *        [_]        [_]         [2]          [3]        [4]
+         *  |___________|___________|___________|______1____|___________|
+         *                         240[e]->    260->       280->       299
+         *
+         * Present
+         *       [0]         [1][b]      [2][g]      [3]         [4]
+         *  |_____1_____|_____1_____|___________|___________|___________|
+         * 300[a]->    320->       340[d]->    360->       380->       399
+         *
+         * Future
+         *       [0]         [_]         [_]         [_]         [_]
+         *  |_____0_____|___________|___________|___________|___________|
+         * 400[c]->    420[f]->
+         *
+         */
+        protected final long resolution;
+        protected final long duration;
+        protected final long[] buckets;
+
+        // The start time of buckets[0]. bucket(t + (i * duration)) is the same for all i. startOfEpoch
+        protected long startOfEpoch;
+        protected int curBucket = 0;
+
+        /**
+         * Create a Counter that covers duration seconds at the given resolution.  Duration must be divisible by resolution.
+         */
+        public Counter(long resolution, long duration) {
+            if (resolution <= 0) {
+                throw new IllegalArgumentException("resolution [" + resolution + "] must be greater than zero");
+            } else if (duration <= 0) {
+                throw new IllegalArgumentException("duration [" + duration + "] must be greater than zero");
+            } else if (duration % resolution != 0) {
+                throw new IllegalArgumentException("duration [" + duration + "] must divisible by resolution [" + resolution + "]");
+            }
+            this.resolution = resolution;
+            this.duration = duration;
+            this.buckets = new long[(int) (duration / resolution)];
+            assert buckets.length > 0;
+        }
+
+        /**
+         * Increment the counter at time {@code t}, expressed in seconds.
+         */
+        public void inc(long t) {
+            if (t < nextBucketStartTime()) {
```

Comment:
Assuming time is monotonic, we are checking that `t` is in the current bucket.  If time is not monotonic, this increments the current bucket anyway to avoid revising a history that may have already been presented to the caller.